File: java/shared/java/org/signal/libsignal/zkgroup/receipts/ReceiptCredentialPresentation.java
//
// Copyright 2020-2021 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

package org.signal.libsignal.zkgroup.receipts;

import java.nio.ByteBuffer;
import org.signal.libsignal.zkgroup.InvalidInputException;
import org.signal.libsignal.zkgroup.internal.ByteArray;
import org.signal.libsignal.internal.Native;

public final class ReceiptCredentialPresentation extends ByteArray {
  public ReceiptCredentialPresentation(byte[] contents) throws InvalidInputException {
    super(contents);
    Native.ReceiptCredentialPresentation_CheckValidContents(contents);
  }

  public long getReceiptExpirationTime() {
    return Native.ReceiptCredentialPresentation_GetReceiptExpirationTime(contents);
  }

  public long getReceiptLevel() {
    return Native.ReceiptCredentialPresentation_GetReceiptLevel(contents);
  }

  public ReceiptSerial getReceiptSerial() {
    byte[] newContents = Native.ReceiptCredentialPresentation_GetReceiptSerial(contents);

    try {
      return new ReceiptSerial(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    }
  }

}

File: node/ts/zkgroup/receipts/ReceiptCredentialPresentation.ts
//
// Copyright 2021 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

import ByteArray from '../internal/ByteArray.js';
import * as Native from '../../Native.js';
import ReceiptSerial from './ReceiptSerial.js';

export default class ReceiptCredentialPresentation extends ByteArray {
  private readonly __type?: never;
  static SIZE = 329;

  constructor(contents: Uint8Array) {
    super(contents, Native.ReceiptCredentialPresentation_CheckValidContents);
  }

  getReceiptExpirationTime(): number {
    return Native.ReceiptCredentialPresentation_GetReceiptExpirationTime(
      this.contents
    );
  }

  getReceiptLevel(): bigint {
    return Native.ReceiptCredentialPresentation_GetReceiptLevel(this.contents);
  }

  getReceiptSerialBytes(): ReceiptSerial {
    return new ReceiptSerial(
      Native.ReceiptCredentialPresentation_GetReceiptSerial(this.contents)
    );
  }
}
